scene/resources/curve.h

```diff
@@ -238,24 +238,37 @@ class Curve3D : public Resource {
 
 	Vector<Point> points;
 
-	struct BakedPoint {
-		real_t ofs = 0.0;
-		Vector3 point;
-	};
+	// unused, remove?
+	//struct BakedPoint {
+	//	real_t ofs = 0.0;
+	//	Vector3 point;
+	//};
 
 	mutable bool baked_cache_dirty = false;
 	mutable PackedVector3Array baked_point_cache;
 	mutable Vector<real_t> baked_tilt_cache;
+	// cache the rotation
 	mutable PackedVector3Array baked_up_vector_cache;
+
 	mutable Vector<real_t> baked_dist_cache;
 	mutable real_t baked_max_ofs = 0.0;
 
 	void mark_dirty();
 
 	void _bake() const;
 
+	struct Interval {
+		int idx;
+		real_t frac;
+	};
+	Interval _find_interval(real_t p_offset) const;
+	Vector3 _sample_baked(Interval p_interval, bool p_cubic) const;
+	real_t _sample_baked_tilt(Interval p_interval) const;
+	Basis _sample_baked_basis(Interval p_interval, bool p_apply_tilt = false) const;
```

Comment:
`sample_baked_basis()` doesn't make sense, what basis is it?

The correct way is to implement `sample_baked_forward_vector()` to `Cureve3D`, then in Path3D call it and `sample_baked_up_vector()`, `sample_baked()` to calculate the transform for the rotation.

By the way, it would be better to rename `sample_baked()` to `sample_baked_position()`. Probably we should do it in sepalate PR.

Reply:
> sample_baked_basis() doesn't make sense, what basis is it?

The return Basis contains the forward, up, and side vector together. This function is essential for **smooth** interpolation of vectors between two neighboring baked point. sample_baked_up_vector is actually implemented by calling `_sample_baked_basis`. If you think the name is not informative, we can changed, as it is only an internal method.

Reply:
As I have said many times, rotation is meaningless to `Curve3D`.

Also, you are not understanding `sample_baked` correctly.

If you want to implement `sample_baked_basis()`, you need to cache the rotation in `Vector<Basis>`, just like tilt and up_vector are stored in an array. Since your current code is performing procedural calculations for the rotation, so `sample_baked` does not make sense.

Also, baking when there are multiple `RotationMode`s means that a `Vector<Vector<Basis>>` with as many elements as there are `RotationMode`s is needed. Having this in `Curve3D` is too evil.

---

First, you need to implement `baked_forward_vector_cache` and `sample_baked_forward_vector()` correctly, referring to the `baked_up_vector_cache` implementation.

Rotation is meaningless in `Curve3D`, so call `sample_baked_/up_vector/forward_vector/position()` in `Path3D` to create `Transform3D`. Then, you can create rotations in `Path3D` as you wish. If necessary, you can cache a `Vector<Transform3D>` within `Path3D`.

Reply:
> rotation is meaningless to Curve3D

I'd like to make sure we are on the same page. 

I understand the `rotation` on a Curve3D as

> T(foward) is the unit vector [tangent](https://en.wikipedia.org/wiki/Tangent_vector) to the curve, pointing in the direction of motion.
> N(side) is the [normal](https://en.wikipedia.org/wiki/Normal_vector) unit vector, the derivative of T with respect to the [arclength parameter](https://en.wikipedia.org/wiki/Rectifiable_path) of the curve, divided by its length.
> B(up) is the binormal unit vector, the [cross product](https://en.wikipedia.org/wiki/Cross_product) of T and N.

if `rotation` means something else to you, what do you call the TNB triple?

if your objection to the `sample_baked_rotation` is based on the word `rotation`, is it possible to accept it after some rename?

Reply:
> T(foward):
> T(foward) is the unit vector [tangent](https://en.wikipedia.org/wiki/Tangent_vector) to the curve, pointing in the direction of motion.
> N(side) is the [normal](https://en.wikipedia.org/wiki/Normal_vector) unit vector, the derivative of T with respect to the [arclength parameter](https://en.wikipedia.org/wiki/Rectifiable_path) of the curve, divided by its length.
> B(up) is the binormal unit vector, the [cross product](https://en.wikipedia.org/wiki/Cross_product) of T and N.

A Godot Engine Basis() is 3 vectors of AXIS_X, AXIS_Y, AXIS_Z.

Maybe use a Vector<Vector3> of size 3 or call TNB something else? 

Also I have doubts that your baked mesh is smooth in the sense of a rollercoaster ride.

[solver to work for G1 splines. It's easier to solve for G2 continuity.](https://twitter.com/FreyaHolmer/status/1583950500650590209?s=20&t=5ZKBo_uYpjHyMVkc0J8PHA)

Reply:
I don't think we need to have it perfectly smooth (or if it turns out to be necessary it can be done in PathFollow because I can see a ton of uses that don't need perfectly smooth)

+1 to exposing Tangent Normal Binormal/Up, whatever the naming for the last one is, I need the first two urgently8 for my non-PathFollow cars.

Reply:
I don't think we need to have it perfectly smooth (or if it turns out to be necessary it can be done in PathFollow because I can see a ton of uses that don't need perfectly smooth)

+1 to exposing Tangent Normal Binormal/Up, whatever the naming for the last one is, I need the first two urgently* for my non-PathFollow cars.

yes I worked around it for now but it's ugly

Reply:
Rotation by TNB frames is needed for solving path moving, not for drawing like CSGShape, and if there is no use case other than Path3D, then I consider that it is more appropriate to calculate it in Path3D.

Reply:
> A Godot Engine Basis() is 3 vectors of AXIS_X, AXIS_Y, AXIS_Z.
If I understand correct it is the AXIS_X, AIXS_Y, AXIS_Z of a **local space**? For me that sounds perfect for TNB, as TNB defined a local space.
![image](https://user-images.githubusercontent.com/6020486/200121117-f0cc6740-aed0-4647-8341-2362e0a8ca22.png)
(picture found on internet)

Of course I can use `Vector<Vector3>` if reviewers insists.

> Also I have doubts that your baked mesh is smooth in the sense of a rollercoaster ride.
My implementable is definitely not the best, probably not as good as this one. It is base on "C. Dougan, [The Parallel Transport Frame](https://gitea.yiem.net/QianMo/Real-Time-Rendering-4th-Bibliography-Collection/raw/branch/main/Chapter%201-24/[0374]%20[Game%20Programming%20Gems%202%202001]%20The%20Parallel%20Transport%20Frame.pdf), Game Programming Gems 2". It definitely much improved than the old implementation, and solved several old issues(see the first post). 

The key idea of this PR is to allow sample TNB frame from `Curve3D` directly without going through `PathFollow3D` or `Path3D`.  Further improvement of the smooth baking is better left to other PRs.

Reply:
> Rotation by TNB frames is needed for solving path moving, not for drawing like CSGShape, and if there is no use case other than Path3D, then I consider that it is more appropriate to calculate it in Path3D.

Then I think we have a fundamental difference in Philosophy. I prefer keep a feature at the lowest level that makes sense. In this case.

Is your decision final, or is there some way I can appeal. If not I think I will left this PR.

Reply:
@xiongyaohua I am making the decision as a member of the Animation team based on Godot's convention that PR without an actual use case will be dismissed. The final decision will be made by @reduz. Or we might want to discuss it once at the PR Meeting.

You can keep this PR. It will be mentions about the final decision, you can fix this if necessary. Or if you have lost interest for about it, someone else can salvage this PR.

Reply:
Reconsideration, the `up_vector` is also not originally needed for `Curve3D`, but I found that a parameter called `up_vector_enabled` already exists to avoid caching when it is not needed. Likewise, if you add a flag like posture_enabled to `Curve3D`, it would be possible to cache and sample basis with `Vector<Basis>`. This should allow for the correct implementation of `sample_baked_basis()`.

Then, to avoid the need to explicitly write `c->set_posture_enabled(false);` in other classes that use `Curve3D`, such as `CSGShape`, it may be better to use a default value of false. But I would like to hear what others have to say about this.

Also, I had overlooked that this same fix had already been merged into `Curve2D`. Considering it is approved, `RotationMode` may also be defined within `Curve3D`.

Reply:
> Also, I had overlooked that this same fix had already been merged into Curve2D. Considering it is approved, RotationMode may also be able to be defined within Curve3D.

That's a good point.

Reply:
> Reconsideration

I am really glad we come to some consensus, as to allow `Curve3D` to be sampled without going thought other layer is the raison d'être of this PR. I will try my best to follow through.

## What `Rotation` is?
I reviewed all our previous discussion, and feel when we use the word `rotation`, we have different things in mind. And that cause unnecessary confusion. To facilitate further discussion,  we need to clarify some terms.

To me `rotation` have two parts:

- The first part is the intrinsic `rotation frame`, like the Frenet(TNB) frame I mentioned before. This part is mathematically well defined, you **can't change it** without inventing new kind of mathematical concepts(like tangent, and binormal). In the following discussion, I will call it `rotation frame` or `frame`.
- The second part is the `artistic control`, like locking X, Y, Z axes, setting up or target directions, etc. This part can be **changed by the user at will**. For example with `RotationMode` parameter.

## My Position
> However, enum RotationMode should not be defined in the Curve.

So when you propose to move the `RotationMode` to `PathFollow3D` , I am actually agreeing with you. I also think the `artistic control` is not an intrinsic part of a `Curve3D`. So it is best defined and implemented on `PathFollow3D`.

> I looked at the code a little more closely. So, sample_baked_basis() doesn't make sense, what "basis" is it in Curve3D?

But when you say this, I thought you imply that `frame` is not meaningful on `Curve3D`, which I cannot agree, as the `frame` is mathematically well defined on curve.

So my position in one sentence: `rotation frame` on `Curve3D`, `artistic control` on `PathFollow3D`. Can we agree on this?

Reply:
> I will call it rotation frame or frame.

"Frame" is not recommended because it conflicts too much with other terms in Godot. Considering the fact that `up_vector` is not called `B_vector` as well, it would be more appropriate to use the meaning by usage. TNB frame / moving frame is used for posture control, so I prefer to use the "posture". Or, it could use "rotation" just in this use case.

> So it is best defined and implemented on PathFollow3D.

The problem is that similar implementations have already been merged in `Curve2D`, so it will not be consistent. In `Curve2D`, `Curve2D:: sample_baked_with_rotation` receives a value called "ahead", which is unrelated to `Curve2D`. If `sample_baked_with_rotation` is defined in `Curve3D`, it is not unnatural for Curve3D to have enum `RotationMode`.

Reply:
> so I prefer to use the "posture". Or, it could use "rotation" just in this use case

Fair enough. Then I use `posture` in our discussion. For API I will keep using "rotation".

> Curve2D::sample_baked_with_rotation receives a value called "ahead", which is unrelated to Curve2D

This "ahead" parameter is not for `artistic control`. It is a hack added to patch non smooth loop, more like "Curve3D::loop_smoothing_enabled"(removed, following your suggestion). It is inherit from old implementation, and no longer necessary. I think I will propose to remove it in follow up PR. Then we will have symmetrical API for 2D/3D.

>  if you add a flag like posture_enabled to Curve3D, it would be possible to cache and sample basis with Vector<Basis>. This should allow for the correct implementation of sample_baked_basis()

Only the up vectors are cached is a conscious decision. In the initial baking implementation I actually cached the whole `posture` ,

https://github.com/xiongyaohua/godot/blob/17577571f6deee55387d6f775006342c1884311f/scene/resources/curve.h#L251-L253

Then I realized I actually don't need to. Because the up(N) vector is the most expensive to get, while the forward(T) and sideway(B) calculation is quite cheap (a vector minus for T, a cross product for B). 

So I decide to only cache up vector, and calculate the other two on the fly. This way we save 2 Vector3 (24 byte) per baked points(normally in hundreds or thousands).

Reply:
Indeed, I found that `PathFollow2D` has `bool rotates`, it would be right to put the `RotationMode` into `PathFollow3D`. Then, I suggest that the `RotationMode` should not be passed to `Curve3D`, but instead the `PathFollow3D` side should adjust for the `Basis` (in `Transform3D`) received from `Curve3D::sample_baked_with_rotation()` depends on the `RotationMode`.

However, if this rotation correction is performed in `PathFollow3D`, it will be difficult to get the path including the posture, which is the expected behavior by @Zireael07 without `PathFollow3D`, so it may be better to implement the rotation correction in `Path3D`.

Reply:
> so it may be better to implement the rotation correction in Path3D like Path3D::get_transform_from_curve(float p_interval, PathFollow3D::RotationMode p_rotation_mode)

I feel hesitant to add logic to `Path3D`. The reason is that, currently its only responsibility is an empty shell that hold the resource `Curve3D` into the scene tree (`debug_instance` and `debug_mesh` notwithstanding).

https://github.com/godotengine/godot/blob/13f1d8096032bf572d5071c5bf9f61de0a055fed/scene/3d/path_3d.h#L37-L45

Adding the `artistic control` logic on `Path3D` will destroy this simplicity and create insymmetry with the 2D counterpart, creating problems in future.

> get the transform with the corrected posture, which is the wanted behavior by @Zireael07

From the previous posts, I get the feeling that @Zireael07 have a very similar use case as my own (If I am wrong, feel free correct me). We mainly need the `rotation frame`(posture) from the curve, and need no or very simple `artistic control`. Then we can just do the necessary rotation in our own code. It's quite simple.

For example, to lock rotation to the up vector
```c++
Vector3 euler = frame.get_euler_normalized(EulerOrder::YXZ);
euler[0] = 0;
euler[2] = 0;
frame = Basis::from_euler(euler, EulerOrder::YXZ);
```
(In c++, can be directly translated into GDScript)

So I suggest we keep the `artistic control` on `PathFollow3D`, which make the life much easier for people who want to implement more sophisticated `artistic control` algorithms like constraints and drivers.

If @Zireael07 do need full blown `artistic control` or If you insist we must do `artistic control` on `Path3D`, I think I can live with it.

Reply:
That static function is no problem to implement in PathFollow3D. Then use bind_static_method in PathFollow3D::_bind_methods().

Reply:
Can you make it a static function?

Reply:
Sure

Reply:
> That static function is no problem to implement in PathFollow3D

my bad, I mistaken for ‘Path3D’

>  Then use bind_static_method in PathFollow3D::_bind_methods().

thanks. It's done

Reply:
The name of the function `_sample_baked_basis()` is still strange if you are not baking the basis. It should be renamed something like `_get_posture()`, `_make_posture()` or `_sample_posture()`. It is not recommended to use `_get_rotation()` or `_get_basis()` as they may be confused with `Node3D` properties

Reply:
done

Reply:
Since the posture is not baked, `_sample_posture` is appropriate instead of `_sample_baked_posture`.

Reply:
done